apps/apps.c

```diff
@@ -2716,6 +2716,125 @@ BIO *bio_open_owner(const char *filename, int format, int private)
     return NULL;
 }
 
+/*
+ * Open a file for writing, owner-read-only. The file is written in
+ * a 'write-to-temp-file-then-rename' basis.
+ */
+#define RND_SIZE 16
+
+static const BIO_METHOD *BIO_f_tempfile();
+
+void make_temp_filename(const char *orig, char *filename)
+{
+    int i, size = RND_SIZE;
+    char *p;
+    char set[] = "0123456789abcdefghijklmnopqrstuvwxyz"
+                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    strcpy(filename, orig);
+    p = filename + strlen(orig);
+    *p++ = '.';
+
+    srand(time(NULL));
+    while (size--) {
+        i = rand() % (sizeof(set) - 1);
+        *p++ = set[i];
+    }
```

Comment:
With this use of srand() and rand(), two openssl commands started within the same second and with the same original output file will trample all over each other, yeah?

We *should* use `mkstemp()`, but I know the issue with that one, you can't be sure that it will end up on the same device as the intended final output file, and therefore risk a `rename()` failure at the end :-/

So it seems that what you need to do is to have the file creation logic in here as well, and make sure to use both `O_CREAT` and `O_EXCL` (thereby ensuring that the `open()` call only succeeds if the file didn't already exist), and loop until you don't get an error.

Reply:
> With this use of srand() and rand(), two openssl commands started within the same second and with the same original output file will trample all over each other, yeah?

Yes. But the probability will be nearly zero I think.

> We should use mkstemp(), but I know the issue with that one, you can't be sure that it will end up on the same device as the intended final output file, and therefore risk a rename() failure at the end :-/

I don't quite understand this. Since `mkstemp()` requires a `template` parameter, if we pass the `template` as something like `original_filename.XXXXXXXXXXXXXXXX`, and because the original_filename has path in it, so the temp file created by mkstemp() will be sure on the same device with the original file, right?

(And we can also just use `mktemp()` to obtain a temp filename only instead of get a `fd` of the temp file, thus we ca n control the creation logic.)

Reply:
Ignore `mkstemp`, I just found out that it's not implemented everywhere (for example in VC).

I would *still* urge you to make sure that you don't create an already existing file, i.e. if you want to reimplement something, reimplement `mkstemp` rather than `mktemp`

Reply:
OK

Reply:
I'm with @levitte here.  It should be doing something more like `mkstemp`.  perhaps move the FD open code from below into this function and put the name generation in a loop (with a small iteration limit).  Return both the file name and the FD.

This definitely can be a problem, there were more than a few attacks using this vector from at least the mid 1980's.  An attacker can often force an overwrite that does something interesting.  Symbolic links are useful.